lychee-lib/src/client.rs

```diff
@@ -646,29 +662,25 @@ impl Client {
     }
 
     /// Check a URI using [reqwest](https://github.com/seanmonstar/reqwest).
-    async fn check_default(&self, uri: &Uri, credentials: &Option<BasicAuthCredentials>) -> Status {
-        let request = match credentials {
-            Some(credentials) => self
-                .reqwest_client
-                .request(self.method.clone(), uri.as_str())
-                .header(AUTHORIZATION, credentials.to_authorization().0.encode())
-                .build(),
-            None => self
-                .reqwest_client
-                .request(self.method.clone(), uri.as_str())
-                .build(),
-        };
+    async fn check_default(&self, uri: &Uri, request_chain: &mut RequestChain) -> Status {
+        let request = self
+            .reqwest_client
+            .request(self.method.clone(), uri.as_str())
+            .build();
 
         let request = match request {
             Ok(r) => r,
             Err(e) => return e.into(),
         };
 
-        let request = self.quirks.apply(request);
+        let result = request_chain.traverse(request);
 
-        match self.reqwest_client.execute(request).await {
-            Ok(ref response) => Status::new(response, self.accepted.clone()),
-            Err(e) => e.into(),
+        match result {
+            EarlyExit(status) => status,
+            Chained(r) => match self.reqwest_client.execute(r).await {
+                Ok(ref response) => Status::new(response, self.accepted.clone()),
+                Err(e) => e.into(),
+            },
```

Comment:
If we had two more states, `ChainResult::Status` and `ChainResult::Err`, could this part also be a part of the `request_chain`?

This way, we might be able to merge the `request_chain` and `response_chain` into one?
Probably there's a reason why we don't want to do this. Just thinking out loud.

Reply:
`reqwest_client.execute` is now part of the request chain in checker.rs
For simplicity I won't add a response chain to this PR if this is fine. But using a separate request and response chain makes more sense in my opinion. This keeps things separate and keeps the Chain type simple.

Plugin users would still want to specify if their plugin/chainable is before or after the request, I think. Making one chain would make this more complicated.